src/kernels/MIOpenCumulativeReduction.hpp

```diff
@@ -0,0 +1,113 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2024 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#ifndef GUARD_KERNELS_MIOPEN_CUMULATIVE_REDUCTIONS_HPP
+#define GUARD_KERNELS_MIOPEN_CUMULATIVE_REDUCTIONS_HPP
+
+#include "float_types.h"
+
+enum class CumulativeReductionOp_t
+{
+    Max  = 1,
+    Min  = 2,
+    Sum  = 3,
+    Prod = 4,
+};
+
+#ifndef __HIP_DEVICE_COMPILE__
+static_assert(MIOPEN_CUM_MAX == static_cast<int>(CumulativeReductionOp_t::Max));
+static_assert(MIOPEN_CUM_MIN == static_cast<int>(CumulativeReductionOp_t::Min));
+static_assert(MIOPEN_CUM_SUM == static_cast<int>(CumulativeReductionOp_t::Sum));
+static_assert(MIOPEN_CUM_PROD == static_cast<int>(CumulativeReductionOp_t::Prod));
+#endif
+
+__device__ inline constexpr void update() {}
+
+template <typename T, typename... Ts>
+__device__ inline constexpr void update(T& a, T b, Ts&... c, Ts... d)
+{
+    a = b;
+    update(c..., d...);
+}
+
+__device__ inline constexpr bool isgreater() { return false; }
+
+template <typename T, typename... Ts>
+__device__ inline constexpr bool isgreater(T& a, T b, Ts&... c, Ts... d)
+{
+    if(a != b)
+        return a > b;
+    return isgreater(c..., d...);
+}
+
+template <typename T, typename... Ts>
+struct reduce_func_base
+{
+    __device__ virtual inline bool isbetter(const T& /*a*/, const T& /*b*/) const { return false; }
+    __device__ virtual inline void combine(T& a, T b) const { a = b; }
```

Comment:
Can we use CRTP instead of virtual functions?

Reply:
Resolved. Please take a look.